### PR TITLE
roberta

### PR DESCRIPTION
**Pull request by 8188zq**
这是nlp预训练模型roberta的pr

### CODE REVIEW

**Comment by rentainhe** [comment]
有参考的Pytorch相关的Repo吗，可以给个Link吗

**Comment by 8188zq** [comment]
> 有参考的Pytorch相关的Repo吗，可以给个Link吗

你好，有的。在readme中有参考transformer的roberta模型的链接：This repo is based on https://github.com/huggingface/transformers/tree/master/src/transformers/models/roberta，然后SST-2这个下游任务的话很简单并没有参考，它是一个用来检验模型正确性的Task.

**Comment by rentainhe** [comment]
> > 有参考的Pytorch相关的Repo吗，可以给个Link吗
> 
> 你好，有的。在readme中有参考transformer的roberta模型的链接：This repo is based on [https://github.com/huggingface/transformers/tree/master/src/transformers/models/roberta，然后SST-2这个下游任务的话很简单并没有参考，它是一个用来检验模型正确性的Task](https://github.com/huggingface/transformers/tree/master/src/transformers/models/roberta%EF%BC%8C%E7%84%B6%E5%90%8ESST-2%E8%BF%99%E4%B8%AA%E4%B8%8B%E6%B8%B8%E4%BB%BB%E5%8A%A1%E7%9A%84%E8%AF%9D%E5%BE%88%E7%AE%80%E5%8D%95%E5%B9%B6%E6%B2%A1%E6%9C%89%E5%8F%82%E8%80%83%EF%BC%8C%E5%AE%83%E6%98%AF%E4%B8%80%E4%B8%AA%E7%94%A8%E6%9D%A5%E6%A3%80%E9%AA%8C%E6%A8%A1%E5%9E%8B%E6%AD%A3%E7%A1%AE%E6%80%A7%E7%9A%84Task).

好的，get，我这边会看一下

**Comment by 8188zq** [comment]
> > > 有参考的Pytorch相关的Repo吗，可以给个Link吗
> > 
> > 
> > 你好，有的。在readme中有参考transformer的roberta模型的链接：This repo is based on [https://github.com/huggingface/transformers/tree/master/src/transformers/models/roberta，然后SST-2这个下游任务的话很简单并没有参考，它是一个用来检验模型正确性的Task](https://github.com/huggingface/transformers/tree/master/src/transformers/models/roberta%EF%BC%8C%E7%84%B6%E5%90%8ESST-2%E8%BF%99%E4%B8%AA%E4%B8%8B%E6%B8%B8%E4%BB%BB%E5%8A%A1%E7%9A%84%E8%AF%9D%E5%BE%88%E7%AE%80%E5%8D%95%E5%B9%B6%E6%B2%A1%E6%9C%89%E5%8F%82%E8%80%83%EF%BC%8C%E5%AE%83%E6%98%AF%E4%B8%80%E4%B8%AA%E7%94%A8%E6%9D%A5%E6%A3%80%E9%AA%8C%E6%A8%A1%E5%9E%8B%E6%AD%A3%E7%A1%AE%E6%80%A7%E7%9A%84Task).
> 
> 好的，get，我这边会看一下

嗯嗯好的，麻烦您了

**Comment by rentainhe** [comment]
记录一下BERT和RoBERTa之间的细节差异以及需要check的重点：
BERT原文: [https://arxiv.org/abs/1810.04805](https://arxiv.org/abs/1810.04805)
RoBERTa原文: [https://arxiv.org/abs/1907.11692](https://arxiv.org/abs/1907.11692)

BERT:
- 两个Segment使用[SEP]拼接
- 纯Transformer Encoder结构
- 学习目标：
  - MLM: 15%的token被Mask，在15%的Mask Token里有80%被替换为[Mask]标记，10%被随机替换为其他token，10%保持不变。
  - NSP，判断第二个segment是否为第一个的后续段落，随机采样出50%的是或者不是
- Optimization:
  - Adam: {beta1=0.9, beta2=0.999, epsilon=1e-6, L2 weight decay=0.01}
  - learning rate: Linear Warmup Decay: {Warmup Steps=10000, lr_base=1e-4}
  - dropout=0.1
  - GELU
  - Total Steps: 1M
  - mini-batch: 256
  - input token length: 512
- Data
  - BookCorpus + English Wiki = 16GB

RoBERTa:
- [ ] Dynamic Masking策略
- [ ] 移除NSP，换成连续的Full-Sentence或者Doc-Sentence
- [ ] Large-Batch，batch size=2k，total steps=125k
- [ ] 增加新的预训练数据集CC-NEWS，16G文本语料增大到160G

File: roberta/roberta/models/dev_ops.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+import oneflow as flow
+
+
+class LayerNorm(flow.nn.Module):
+
+    def __init__(self, features, eps=1e-6):
+        super(LayerNorm, self).__init__()
+        self.eps = eps
+        self.a_2 = flow.nn.Parameter(flow.Tensor(
```
Comment (rentainhe) [inline roberta/roberta/models/dev_ops.py:9]: 在载入权重的时候如果这里的weight和bias定义为`a_2`和`b_2`可以和预训练权重的state_dict对应上吗，刚刚用torch验证的时候打印出来的是 **'a_2' : tensor()**，建议改成`self.weight`和`self.bias`与内部定义对齐

File: roberta/roberta/infer.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    attention_mask = tokenizer(args.text)['attention_mask']
+    input_ids = flow.tensor(input_ids, dtype=flow.int32).reshape(1,-1).to('cuda')
+    attention_mask = flow.tensor(attention_mask, dtype=flow.int32).reshape(1,-1).to('cuda')
+    model = GlueRoBERTa(args.pretrain_dir, args.kwargs_path,
+                        args.roberta_hidden_size, args.n_classes, args.is_train).to('cuda')
+    model.load_state_dict(flow.load(args.model_load_dir))
+    model.eval()
+    output = model(input_ids,attention_mask)
+    label = flow.argmax(output)
+    # label, logits = predict(model, text)
```
Comment (rentainhe) [inline roberta/roberta/infer.py:23]: 可以删除一下不必要的注释

File: roberta/roberta/infer.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    parser.add_argument('--kwargs_path', type=str,
+                        default='./roberta_pretrain_oneflow/roberta-base/roberta-base.json')
+    parser.add_argument('--pretrain_dir', type=str,
+                        default='./roberta_pretrain_oneflow/roberta-base')
+    parser.add_argument('--text', type=str,
+                        default="this is junk food cinema at its greasiest .")
+
+                
+    args = parser.parse_args()
+    args.n_classes = 2
```
Comment (rentainhe) [inline roberta/roberta/infer.py:40]: 个人想法是根据不同task整理一个统一的config，虽然目前暂时只有SST-2上的验证，后续最好再添加一下其他task的验证任务
```python
def infer_config(args):
    if args.task = 'SST-2':
        args.n_classes = 2
        args.roberta_hidden_size = 768
        args.is_train = False
```

File: roberta/roberta/train_flow.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    model = GlueRoBERTa(args.pretrain_dir, args.kwargs_path,
+                        args.roberta_hidden_size, args.n_classes, args.is_train).to('cuda')
+    criterion = oneflow.nn.CrossEntropyLoss().to('cuda')
+    of_adam = oneflow.optim.Adam(model.parameters(), args.lr)
+
+    time_map['t1'] = time.time()
+    print("beging to load data...")
+    start = time.time()
+    input_ids, attention_mask, labels = read_data('train')
+    train_data = myDataset(input_ids, attention_mask, labels)
```
Comment (rentainhe) [inline roberta/roberta/train_flow.py:30]: 可以的话还是改一下`myDataset`的名字吧

File: roberta/roberta/train_flow.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    parser.add_argument('--n_epochs', type=int, default=30)
+    parser.add_argument('--kwargs_path', type=str,
+                        default='./roberta_pretrain_oneflow/roberta-base/roberta-base.json')
+    parser.add_argument('--pretrain_dir', type=str,
+                        default='./roberta_pretrain_oneflow/roberta-base')
+    parser.add_argument('--model_save_dir', type=str,
+                        default='./pretrain_model_SST-2')
+
+    args = parser.parse_args()
+    args.n_classes = 2
```
Comment (rentainhe) [inline roberta/roberta/train_flow.py:104]: 同上，根据train的task定制一个统一的train_config

**Review by rentainhe — COMMENTED** [review]
Partly Review Comment：

该PR主要实现的内容：
- [ ] 基于Transformers框架抽取出RoBERTa模型的必要组件
- [ ] 载入RoBERTa预训练权重在SST-2数据集上验证准确率
- [ ] 载入RoBERTa在下游任务上fine-tune

实现相对复杂，可以进行部分简化，参考验证的下游任务可以增加，不止SST-2，最好有条件的情况下把原文里提到的下游任务都验证一下（可能需要修改一下output，定制一下不同任务的config：
- [ ] MNLI
- [ ] QNLI
- [ ] QQP
- [ ] RTE
- [ ] SST
- [ ] MRPC
- [ ] CoLA
- [ ] STS
- [ ] WNLI

File: roberta/roberta/train_flow.py (hunk 5 above)
Comment (8188zq) [reply roberta/roberta/train_flow.py]: 已经在新的版本中改进

File: roberta/roberta/train_flow.py (hunk 4 above)
Comment (8188zq) [reply roberta/roberta/train_flow.py]: 已经在新的版本中改好

File: roberta/roberta/infer.py (hunk 3 above)
Comment (8188zq) [reply roberta/roberta/infer.py]: 已经在新的版本中改好

File: roberta/roberta/infer.py (hunk 2 above)
Comment (8188zq) [reply roberta/roberta/infer.py]: 已经在新的版本中改好

File: roberta/roberta/models/dev_ops.py (hunk 1 above)
Comment (8188zq) [reply roberta/roberta/models/dev_ops.py:9]: 已经在新的版本中改好

Comment (8188zq) [reply roberta/roberta/models/dev_ops.py:9]: 新增了MNLI ( natural language inference task)

**Comment by rentainhe** [comment]
可以按照新版的models更新一下吗
